### PR TITLE
fix: only care about ecosystem suffix if present in both ecosystems when determining equality

### LINKED ISSUE #769: Missing vulnerabilities for debian purls for --experimental-local-db
I tried running --format json --sbom bom.json --experimental-local-db ~/database and it missed all the purls that has 
pkg:deb/debian but I get hits for --format json --sbom bom.json

### PR DESCRIPTION
**Pull request by G-Rath**
This changes how we compare ecosystems so that the suffix is only considered when it is present for _both_ ecosystems being compared, since we can't reliably extract that.

Resolves #769

### CODE REVIEW

File: cmd/osv-scanner/__snapshots__/main_test.snap
Comment (G-Rath) [inline cmd/osv-scanner/__snapshots__/main_test.snap:144]: the fact that the offline output has vulnerabilities whereas the online one doesn't indicates there's a potential bug in the api's Alpine comparator 🤔

Comment (another-rex) [reply cmd/osv-scanner/__snapshots__/main_test.snap:144]: Yes there is a bug in alpine version enumeration, just fixed this week in: https://github.com/google/osv.dev/pull/2241 but not in production yet.

Comment (G-Rath) [reply cmd/osv-scanner/__snapshots__/main_test.snap:144]: Cool - I'd seen a few of those go around this/last week, so figured it might have been known 😅

**Review by another-rex — COMMENTED** [review]
LGTM, thanks!. Going to wait till the next OSV.dev release (with the alpine enumeration fix) before merging this, so we'll have the correct snapshots.

**Comment by G-Rath** [comment]
Turns out we're not accurately sorting the table output at least for local - I'll tackle that somewhere...

**Review by hogo6002 — APPROVED** [review]
LGTM

**Review by cuixq — APPROVED** [review]

**Comment by G-Rath** [comment]
I cannot for the life of me replicate that sorting difference locally 😕

**Comment by hogo6002** [comment]
> I cannot for the life of me replicate that sorting difference locally 😕

I tried running  `update_snaps` for this branch locally, and it did change the order, but it's still a little different from the [Ubuntu results](https://github.com/google/osv-scanner/actions/runs/9488640750/job/26148082254?pr=1007). 
image

**Comment by G-Rath** [comment]
Yeah, it seems that every couple of hours the order changes - I'm not sure if its coinciding with something on the osv.dev side like work being landed or if it's just some clock-shift based randomness, but it seems to be "hours" based rather than "seconds" based...

**Comment by hogo6002** [comment]
> Yeah, it seems that every couple of hours the order changes - I'm not sure if its coinciding with something on the osv.dev side like work being landed or if it's just some clock-shift based randomness, but it seems to be "hours" based rather than "seconds" based...

Yeah I see what you mean. I just tested again and the result shows different.

**Comment by hogo6002** [comment]
I had a look with @michaelkedar  on this issue. We noticed that the result changes after the export cron job publishes a new `all.zip`. We think this is because the export runs in parallel (https://github.com/google/osv.dev/pull/2167), which may cause the order to be slightly different each time. To solve this, I guess we need to either sort the `all.zip` on the OSV.dev side or order all vulnerabilities alphabetically on the OSV-Scanner side. What do you think @another-rex

**Comment by G-Rath** [comment]
oohh good find - fwiw locally I've been playing around with sorting the table rows within packages by their url though that got blocked by it being an `interface{}`; I've not gotten back to it yet so might be an easy fix

**Comment by another-rex** [comment]
Hmm... is this behaviour new? All.zip is built concurrently, but that has always been the case if I understand correctly.

**Comment by hogo6002** [comment]
> Hmm... is this behaviour new? All.zip is built concurrently, but that has always been the case if I understand correctly.

I don't think a new behaviour. Even though all.zip is built concurrently, the files are usually modified in alphabetical order, just a few are out of place. That's probably why most of our tests are passing. But this PR has a bunch of vulnerabilities with adjacency IDs (like 9840, 9841, 9842, and 9843), which is probably why it's failing. But I am not 100% sure if this is the issue as the modified order doesn't really align with the scanner output order.
image

**Comment by another-rex** [comment]
I think this just needs a flag change in the test to match the new download-offline-db flag and should be good to merge.

**Pull request merged**
